src/include/mpir_func.h

```diff
@@ -21,51 +21,22 @@
+#define MPIR_FUNC_EXIT_(level)  MPL_DBG_MSG_S(MPL_DBG_ROUTINE_EXIT, level, "Leaving %s", __func__)
 #elif defined(MPICH_DEBUG_MEMARENA)
-#define MPIR_FUNC_ENTER(a) MPL_trvalid("Entering " #a)
-#endif
-
-#if defined(MPL_USE_DBG_LOGGING)
-#define MPIR_FUNC_EXIT(a) MPL_DBG_MSG(MPL_DBG_ROUTINE_EXIT,TYPICAL,"Leaving "#a)
-#elif defined(MPICH_DEBUG_MEMARENA)
-#define MPIR_FUNC_EXIT(a) MPL_trvalid("Leaving " #a)
+#define MPIR_FUNC_ENTER MPL_trvalid2("line %d - entering %s\n", __LINE__, __func__)
+#define MPIR_FUNC_EXIT MPL_trvalid2("line %d - leaving %s\n", __LINE__, __func__)
 #endif
 
 /* state declaration macros */
 #if defined(MPL_USE_DBG_LOGGING) || defined(MPICH_DEBUG_MEMARENA)
 
-#define MPIR_FUNC_TERSE_STATE_DECL(a)
-#define MPIR_FUNC_TERSE_INIT_STATE_DECL(a)
-#define MPIR_FUNC_TERSE_FINALIZE_STATE_DECL(a)
-#define MPIR_FUNC_TERSE_ENTER(a)             MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_EXIT(a)              MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_PT2PT_ENTER(a)       MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_PT2PT_EXIT(a)        MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_PT2PT_ENTER_FRONT(a) MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_PT2PT_EXIT_FRONT(a)  MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_PT2PT_ENTER_BACK(a)  MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_PT2PT_ENTER_BOTH(a)  MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_PT2PT_EXIT_BACK(a)   MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_PT2PT_EXIT_BOTH(a)   MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_COLL_ENTER(a)        MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_COLL_EXIT(a)         MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_RMA_ENTER(a)         MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_RMA_EXIT(a)          MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_REQUEST_ENTER(a)	 MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_REQUEST_EXIT(a)		 MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_INIT_ENTER(a)        MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_INIT_EXIT(a)         MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_TERSE_FINALIZE_ENTER(a)    MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_TERSE_FINALIZE_EXIT(a)     MPIR_FUNC_EXIT(a)
-
-#define MPIR_FUNC_VERBOSE_STATE_DECL(a)
-#define MPIR_FUNC_VERBOSE_ENTER(a)           MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_VERBOSE_EXIT(a)            MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_VERBOSE_PT2PT_ENTER(a)     MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_VERBOSE_PT2PT_EXIT(a)      MPIR_FUNC_EXIT(a)
-#define MPIR_FUNC_VERBOSE_RMA_ENTER(a)       MPIR_FUNC_ENTER(a)
-#define MPIR_FUNC_VERBOSE_RMA_EXIT(a)        MPIR_FUNC_EXIT(a)
+#define MPIR_FUNC_ENTER                   MPIR_FUNC_ENTER_(TYPICAL)
+#define MPIR_FUNC_EXIT                    MPIR_FUNC_EXIT_(TYPICAL)
+#define MPIR_FUNC_TERSE_ENTER             MPIR_FUNC_ENTER_(TERSE)
+#define MPIR_FUNC_TERSE_EXIT              MPIR_FUNC_EXIT_(TERSE)
+#define MPIR_FUNC_VERBOSE_ENTER           MPIR_FUNC_ENTER_(VERBOSE)
+#define MPIR_FUNC_VERBOSE_EXIT            MPIR_FUNC_EXIT_(VERBOSE)
```

Comment:
Is the difference between these three versions meaningful? Do people use all three or know when to use all three? Would it make more sense to just have enter and exit and be done?

Reply:
We have different developers/users and I find (in a hard way) that we all have different preferences 😄 

Users, for example, may not be interested in internal functions, so having the top level functions and internal functions logged at different level can serve that separation. Then even when we want to see internal functions logged, we may desire to put those much smaller non-interesting functions to the verbose category to ease the log parsing. So I think have the flexibility of a few logging levels can come in handy.

Yes, by large we just default to use `MPIR_FUNC_ENTER/EXIT` and be done.

Reply:
I'm not sure how many users use the logs here aside from when we ask them to. Maybe you see something different than I do.

I agree that flexibility is nice, but what I see in practice is that most of the time people forget to put these in at all and when they do , they're very unlikely to use anything other than the default.

The logs are already outrageously verbose in most cases (intentionally) and saving a little bit of space by using one form or another seems like an optimization that's not particularly useful.

Reply:
We had quite a few PRs already just to rename the STATE name. Add boilerplates is one thing, maintaining the names is another hassle. When the names are inconsistent, it becomes misleading and worse than not having a log at all. We constantly run into this issue as we constantly moving code around and rename functions. This PR removes the name maintenance burden, which IMO is the bigger part of the burden.

Reply:
My preference would be to have a single option, but I won't strongly advocate against the existing 3 granularities.